tmt/steps/provision/__init__.py

```diff
@@ -1918,45 +2045,147 @@ def go(self, force: bool = False) -> None:
+
+        def _run_action_phases(phases: list[Action]) -> tuple[list[ActionTask], list[ActionTask]]:
+            queue: PhaseQueue[ProvisionStepData] = PhaseQueue(
+                'finish',
+                self._logger.descend(logger_name=f'{self}.queue'))
+
+            for action in phases:
+                queue.enqueue_action(phase=action)
+
+            all_tasks: list[ActionTask] = []
+            failed_tasks: list[ActionTask] = []
+
+            for outcome in queue.run():
+                assert isinstance(outcome, ActionTask)
+
+                all_tasks.append(outcome)
+
+                if outcome.exc:
+                    outcome.logger.fail(str(outcome.exc))
+
+                    failed_tasks.append(outcome)
+
+                    continue
+
+            return all_tasks, failed_tasks
+
+        # Provisioning phases may be intermixed with actions. To perform all
+        # phases and action in a consistent manner, we will process them in
+        # the order or their `order` key. We will group provisioning phases
+        # not interrupted by action into batches, and run the sequence of
+        # provisioning phases in parallel.
+        all_phases = self.phases(classes=(Action, ProvisionPlugin))
+        all_phases.sort(key=lambda x: x.order)
+
+        all_outcomes: list[Union[ActionTask, ProvisionTask]] = []
+        failed_outcomes: list[Union[ActionTask, ProvisionTask]] = []
+
+        while all_phases:
+            # Start looking for sequences of phases of the same kind. Collect
+            # as many as possible, until hitting a different one
+            phase = all_phases.pop(0)
+
+            if isinstance(phase, Action):
+                action_phases: list[Action] = [phase]
+
+                while all_phases and isinstance(all_phases[0], Action):
+                    action_phases.append(cast(Action, all_phases.pop(0)))
+
+                _all_action_outcomes, _failed_action_outcomes = _run_action_phases(action_phases)
+
+                all_outcomes += _all_action_outcomes
+                failed_outcomes += _failed_action_outcomes
+
+            else:
+                plugin_phases: list[ProvisionPlugin[ProvisionStepData]] = [
+                    phase]  # type: ignore[list-item]
+
+                # ignore[attr-defined]: mypy does not recognize `phase` as `ProvisionPlugin`.
+                if phase._thread_safe:  # type: ignore[attr-defined]
+                    while all_phases:
+                        if not isinstance(all_phases[0], ProvisionPlugin):
+                            break
+
+                        if not all_phases[0]._thread_safe:
+                            break
+
+                        plugin_phases.append(
+                            cast(
+                                ProvisionPlugin[ProvisionStepData],
+                                all_phases.pop(0)))
+
+                _all_plugin_outcomes, _failed_plugin_outcomes = _run_provision_phases(
```

Comment:
Why prefixing these variables with underscores?

Reply:
Seems like a relict, this part used to be slightly different. Removed in db0510dc.